### PR TITLE
feat: add connection string redaction utilities COMPASS-5308

### PR DESCRIPTION
**Pull request by addaleax**

### CODE REVIEW

File: src/redact.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+export function redactConnectionString(uri: string): string {
+  try {
+    // If we can parse the connection string, use the more precise
+    // redaction logic.
+    return (new ConnectionString(uri))
+      .redact()
+      .toString()
+      .replace(/_credentials_/g, '<credentials>');
+  } catch {}
```
Comment (lerouxb) [inline src/redact.ts:36]: Is there no way to identify parse errors? This will catch and ignore any programming error in redactValidConnectionString()

Comment (addaleax) [reply src/redact.ts:36]: I guess we can restrict this to only catching errors from the ConnectionString constructor, yeah 👍

Comment (addaleax) [reply src/redact.ts]: @lerouxb Restricted the try/catch scope here to that – feel free to take a look :)

**Review by lerouxb — APPROVED** [review]
:+1:

**Comment by addaleax** [comment]
I’ve merged this, but I wanted to leave as a comment here that there is ongoing conversation about the exact format of proxy options in the connection string in https://github.com/mongodb/specifications/pull/1103#discussion_r762362073. We _may_ need to re-adjust the logic here if we do decide to change the spec. However, I don’t see any immediate downside to publishing and using this code as-is.

**Pull request merged**